include/internal/event_queue.h

```diff
@@ -0,0 +1,175 @@
+ */
+
+#ifndef OSSL_INTERNAL_EVENT_QUEUE_H
+# define OSSL_INTERNAL_EVENT_QUEUE_H
+# pragma once
+
+# include "internal/priority_queue.h"
+# include "internal/time.h"
+
+/*
+ * Opaque type holding an event.
+ */
+typedef struct ossl_event_s OSSL_EVENT;
+
+DEFINE_PRIORITY_QUEUE_OF(OSSL_EVENT);
+
+/*
+ * Generic callback function type.
+ */
+typedef int ossl_event_callback_fn(OSSL_EVENT *event, void *callback_data);
+
+/*
+ * Opaque type hold a subscription to a type of events
+ */
+typedef struct ossl_event_subscription_s OSSL_EVENT_SUBSCRIPTION;
+
+/*
+ * Public type representing an event queue, the underlying structure being
+ * opaque.
+ */
+typedef struct ossl_event_queue_s OSSL_EVENT_QUEUE;
+
+/*
+ * Public type representing a event queue entry.
+ * It is public so that it can be embedded into other structures,
+ * it should otherwise be treated as opaque.
+ */
+struct ossl_event_s {
+    uint32_t type;              /* What type of event this is               */
+    uint32_t priority;          /* What priority this event has             */
+    OSSL_TIME when;             /* When the event is scheduled to happen    */
+    void *ctx;                  /* User argument passed to call backs       */
+    void *payload;              /* Event specific data of unknown kind      */
+    size_t payload_size;        /* Length (in bytes) of event specific data */
+
+    /* These fields are for internal use only */
+    size_t ref;                 /* ID for this event (internal use only)    */
+    unsigned int flag_dynamic : 1;  /* Malloced or not?                     */
+};
+
+/*
+ * Utility inline function to populate an event structure
+ */
+static ossl_unused ossl_inline
+void ossl_event_set(OSSL_EVENT *event, uint32_t type, uint32_t priority,
+                    OSSL_TIME when, void *ctx,
+                    void *payload, size_t payload_size)
+{
+    event->type = type;
+    event->priority = priority;
+    event->when = when;
+    event->ctx = ctx;
+    event->payload = payload;
+    event->payload_size = payload_size;
+    /*event->ref = 0;*/
+    /*event->flag_dynamic = 0;*/
+}
+
+/*
+ * Utility functions to extract event fields
+ */
+static ossl_unused ossl_inline
+uint32_t ossl_event_get_type(const OSSL_EVENT *event)
+{
+    return event->type;
+}
+
+static ossl_unused ossl_inline
+uint32_t ossl_event_get_priority(const OSSL_EVENT *event)
+{
+    return event->priority;
+}
+
+static ossl_unused ossl_inline
+OSSL_TIME ossl_event_get_when(const OSSL_EVENT *event)
+{
+    return event->when;
+}
+
+static ossl_unused ossl_inline
+void *ossl_event_get0_payload(const OSSL_EVENT *event, size_t *length)
+{
+    if (length != NULL)
+        *length = event->payload_size;
+    return event->payload;
+}
+
+
+/*
+ * Dynamically create and free events
+ */
+OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                           OSSL_TIME when, void *ctx,
+                           void *payload, size_t payload_size);
```

Comment:
`priority` and `when` are scheduling parameters.  As such, I would rather have them passed to the queuing calls, i.e. `ossl_event_queue_add()`.

Where those parameters are *stored*, I don't really care.

Reply:
Er, I don't see `type` as a **scheduling** parameter.  Sure it's a dispatching parameter in so far that you choose subscribers to call based on the type, but I don't see that as **scheduling** per se

Reply:
(see it like this: `priority` and `when` affects where the event ends up in the queue order.  `type` does not)

Reply:
I had a thought over night -> what about a new_and_add call?
The _set call disappears.

Both the new_and_add and the _add call take the full suite of parameters.
I expect that every call to _new will be immediately followed by a call to _add.

Reply:
> I had a thought over night -> what about a new_and_add call? The _set call disappears.

That implies that embedded event structures will no longer be supported, right?

> Both the new_and_add and the _add call take the full suite of parameters. I expect that every call to _new will be immediately followed by a call to _add.

That is, of course, a solution too...  nice escape from our conflict in opinion :wink:

Reply:
I was thinking that new_and_add() creates an event while add() uses an embedded event.

Reply:
I'm a little confused...  how's the embedded event being populated?

Reply:
`_add(queue, &event, type, priority, when, ctx, payload, payload_size);`

Reply:
I would call that `_set_and_add`, to go along with `_new_and_add`